src/metric-engine/src/engine/open.rs

```diff
@@ -58,35 +58,66 @@ impl MetricEngineInner {
                 self.transform_open_physical_region_request(request);
             all_requests.push((metadata_region_id, open_metadata_region_request));
             all_requests.push((data_region_id, open_data_region_request));
-            physical_region_ids.push((region_id, physical_region_options));
-            data_region_ids.insert(data_region_id);
+            physical_region_ids.insert(region_id, physical_region_options);
         }
 
-        let results = self
+        let mut results = self
             .mito
             .handle_batch_open_requests(parallelism, all_requests)
             .await
             .context(BatchOpenMitoRegionSnafu {})?
             .into_iter()
-            .filter(|(region_id, _)| data_region_ids.contains(region_id))
-            .collect::<Vec<_>>();
+            .collect::<HashMap<_, _>>();
 
+        let mut responses = Vec::with_capacity(physical_region_ids.len());
         for (physical_region_id, physical_region_options) in physical_region_ids {
-            let primary_key_encoding = self
-                .mito
-                .get_primary_key_encoding(physical_region_id)
-                .context(PhysicalRegionNotFoundSnafu {
-                    region_id: physical_region_id,
-                })?;
-            self.recover_states(
-                physical_region_id,
-                primary_key_encoding,
-                physical_region_options,
-            )
-            .await?;
+            let metadata_region_id = utils::to_metadata_region_id(physical_region_id);
+            let data_region_id = utils::to_data_region_id(physical_region_id);
+            let metadata_region_result = results.remove(&metadata_region_id);
+            let data_region_result = results.remove(&data_region_id);
+            let response = self
+                .open_physical_region_with_results(
+                    metadata_region_result,
+                    data_region_result,
+                    physical_region_id,
+                    physical_region_options,
+                )
+                .await
+                .map_err(BoxedError::new);
+            responses.push((physical_region_id, response));
         }
 
-        Ok(results)
+        Ok(responses)
+    }
+
+    async fn open_physical_region_with_results(
+        &self,
+        metadata_region_result: Option<std::result::Result<RegionResponse, BoxedError>>,
+        data_region_result: Option<std::result::Result<RegionResponse, BoxedError>>,
```

Comment:
We can handle the `Option<Result<>>` in `handle_batch_open_requests()` so this method can receive `RegionResponse` directly.

Reply:
It's by designed, this allows the function to continue processing remaining regions even if some fail.